### PR TITLE
[Enhancement]aws sdk get/put object support zero copy

### PR DESCRIPTION
**Pull request by zombee0**
## Why I'm doing:
perf_ssb100g_oss

## What I'm doing:
largely reduced memory copy.
perf_aws_zero

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

## Documentation PRs only:

If you are submitting a PR that adds or changes English documentation and have not
included Chinese documentation, then you can check the box to request GPT to translate the
English doc to Chinese. Please ensure to uncheck the **Do not translate** box if translation is needed.
The workflow will generate a new PR with the Chinese translation after this PR is merged.

- [ ] Yes, translate English markdown files with GPT
- [x] Do not translate

### CODE REVIEW

File: be/src/io/s3_zero_copy_iostream.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    S3ZeroCopyIOStream(const char* buf, size_t size) : S3ZeroCopyIOStream(const_cast<char*>(buf), size) {}
+
+    S3ZeroCopyIOStream(const S3ZeroCopyIOStream&) = delete;
+    void operator=(const S3ZeroCopyIOStream&) = delete;
+    S3ZeroCopyIOStream(S3ZeroCopyIOStream&&) = delete;
+    void operator=(S3ZeroCopyIOStream&&) = delete;
+
+    ~S3ZeroCopyIOStream() {
+        // corresponding new in constructor
+        delete rdbuf();
```
Comment (trueeyu) [inline be/src/io/s3_zero_copy_iostream.h:42]: Maybe double free?

Comment (zombee0) [reply be/src/io/s3_zero_copy_iostream.h:42]: no worries, `PreallocatedStreamBuf` won't take ownership of the underlying buffer.

File: be/src/io/s3_input_stream.cpp [hunk 2]
```diff
@@ -41,19 +43,22 @@ StatusOr<int64_t> S3InputStream::read(void* out, int64_t count) {
...
+        return Aws::New<S3ZeroCopyIOStream>(AWS_ALLOCATE_TAG, reinterpret_cast<char*>(out), real_length);
+    });
 
     Aws::S3::Model::GetObjectOutcome outcome = _s3client->GetObject(request);
     if (outcome.IsSuccess()) {
-        Aws::IOStream& body = outcome.GetResult().GetBody();
-        body.read(static_cast<char*>(out), count);
-        _offset += body.gcount();
-        return body.gcount();
+        _offset += real_length;
```
Comment (DorianZheng) [inline be/src/io/s3_input_stream.cpp:60]: Why not reuse `outcome.GetResult().GetBody().gcount()`? Can we make sure it always equal to `real_length`?

**Review by DorianZheng — COMMENTED** [review]
left a comment. rest LGTM

File: be/src/io/s3_input_stream.cpp (hunk 2 above)
Comment (zombee0) [reply be/src/io/s3_input_stream.cpp:60]: iostream's gcount is set after extract operation(like read), we didn't call read at all(we used to call body.read(xxx), and we can set the correct offset even if we had a wrong range #49692)

**Review by trueeyu — DISMISSED** [review]

**Review by trueeyu — APPROVED** [review]

**Review by DorianZheng — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Pull request merged**